/**
 * Resolve any semvers including `*` that weren't satisifed during resolving dependencies.
 *
 * Note that this isn't really optimal yet and probably won't ever be!
 * Please use wide semver ranges!
 *
 * To do:
 *
 *   - resolve all semver ranges before `*`s
 */

(function(
  // Reliable reference to the global object (i.e. window in browsers).
  global,

  // Dummy constructor that we use as the .constructor property for
  // functions that return Generator objects.
  GeneratorFunction,

  // Undefined value, more compressible than void 0.
  undefined
) {
  var hasOwn = Object.prototype.hasOwnProperty;

  if (global.wrapGenerator) {
    return;
  }

  function wrapGenerator(innerFn, self, tryList) {
    return new Generator(innerFn, self || null, tryList || []);
  }

  global.wrapGenerator = wrapGenerator;
  if (typeof exports !== "undefined") {
    exports.wrapGenerator = wrapGenerator;
  }

  var GenStateSuspendedStart = "suspendedStart";
  var GenStateSuspendedYield = "suspendedYield";
  var GenStateExecuting = "executing";
  var GenStateCompleted = "completed";

  // Returning this object from the innerFn has the same effect as
  // breaking out of the dispatch switch statement.
  var ContinueSentinel = {};

  wrapGenerator.mark = function(genFun) {
    genFun.constructor = GeneratorFunction;
    return genFun;
  };

  // Ensure isGeneratorFunction works when Function#name not supported.
  if (GeneratorFunction.name !== "GeneratorFunction") {
    GeneratorFunction.name = "GeneratorFunction";
  }

  wrapGenerator.isGeneratorFunction = function(genFun) {
    var ctor = genFun && genFun.constructor;
    return ctor ? GeneratorFunction.name === ctor.name : false;
  };

  function Generator(innerFn, self, tryList) {
    var generator = this;
    var context = new Context(tryList);
    var state = GenStateSuspendedStart;

    function invoke(method, arg) {
      if (state === GenStateExecuting) {
        throw new Error("Generator is already running");
      }

      if (state === GenStateCompleted) {
        throw new Error("Generator has already finished");
      }

      while (true) {
        var delegate = context.delegate;
        if (delegate) {
          try {
            var info = delegate.generator[method](arg);

            // Delegate generator ran and handled its own exceptions so
            // regardless of what the method was, we continue as if it is
            // "next" with an undefined arg.
            method = "next";
            arg = undefined;

          } catch (uncaught) {
            context.delegate = null;

            // Like returning generator.throw(uncaught), but without the
            // overhead of an extra function call.
            method = "throw";
            arg = uncaught;

            continue;
          }

          if (info.done) {
            context[delegate.resultName] = info.value;
            context.next = delegate.nextLoc;
          } else {
            state = GenStateSuspendedYield;
            return info;
          }

          context.delegate = null;
        }

        if (method === "next") {
          if (state === GenStateSuspendedStart &&
              typeof arg !== "undefined") {
            // https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generatorresume
            throw new TypeError(
              "attempt to send " + JSON.stringify(arg) + " to newborn generator"
            );
          }

          if (state === GenStateSuspendedYield) {
            context.sent = arg;
          } else {
            delete context.sent;
          }

        } else if (method === "throw") {
          if (state === GenStateSuspendedStart) {
            state = GenStateCompleted;
            throw arg;
          }

          if (context.dispatchException(arg)) {
            // If the dispatched exception was caught by a catch block,
            // then let that catch block handle the exception normally.
            method = "next";
            arg = undefined;
          }
        }

        state = GenStateExecuting;

        try {
          var value = innerFn.call(self, context);

          // If an exception is thrown from innerFn, we leave state ===
          // GenStateExecuting and loop back for another invocation.
          state = context.done
            ? GenStateCompleted
            : GenStateSuspendedYield;

          var info = {
            value: value,
            done: context.done
          };

          if (value === ContinueSentinel) {
            if (context.delegate && method === "next") {
              // Deliberately forget the last sent value so that we don't
              // accidentally pass it on to the delegate.
              arg = undefined;
            }
          } else {
            return info;
          }

        } catch (thrown) {
          state = GenStateCompleted;

          if (method === "next") {
            context.dispatchException(thrown);
          } else {
            arg = thrown;
          }
        }
      }
    }

    generator.next = invoke.bind(generator, "next");
    generator.throw = invoke.bind(generator, "throw");
  }

  Generator.prototype.toString = function() {
    return "[object Generator]";
  };

  function pushTryEntry(triple) {
    var entry = { tryLoc: triple[0] };

    if (1 in triple) {
      entry.catchLoc = triple[1];
    }

    if (2 in triple) {
      entry.finallyLoc = triple[2];
    }

    this.tryEntries.push(entry);
  }

  function resetTryEntry(entry, i) {
    var record = entry.completion || {};
    record.type = i === 0 ? "normal" : "return";
    delete record.arg;
    entry.completion = record;
  }

  function Context(tryList) {
    // The root entry object (effectively a try statement without a catch
    // or a finally block) gives us a place to store values thrown from
    // locations where there is no enclosing try statement.
    this.tryEntries = [{ tryLoc: "root" }];
    tryList.forEach(pushTryEntry, this);
    this.reset();
  }

  Context.prototype = {
    constructor: Context,

    reset: function() {
      this.prev = 0;
      this.next = 0;
      this.sent = undefined;
      this.done = false;
      this.delegate = null;

      this.tryEntries.forEach(resetTryEntry);

      // Pre-initialize at least 20 temporary variables to enable hidden
      // class optimizations for simple generators.
      for (var tempIndex = 0, tempName;
           hasOwn.call(this, tempName = "t" + tempIndex) || tempIndex < 20;
           ++tempIndex) {
        this[tempName] = null;
      }
    },

    stop: function() {
      this.done = true;

      var rootEntry = this.tryEntries[0];
      var rootRecord = rootEntry.completion;
      if (rootRecord.type === "throw") {
        throw rootRecord.arg;
      }

      return this.rval;
    },

    keys: function(object) {
      var keys = [];
      for (var key in object) {
        keys.push(key);
      }
      keys.reverse();

      // Rather than returning an object with a next method, we keep
      // things simple and return the next function itself.
      return function next() {
        while (keys.length) {
          var key = keys.pop();
          if (key in object) {
            next.value = key;
            next.done = false;
            return next;
          }
        }

        // To avoid creating an additional object, we just hang the .value
        // and .done properties off the next function object itself. This
        // also ensures that the minifier will not anonymize the function.
        next.done = true;
        return next;
      };
    },

    dispatchException: function(exception) {
      if (this.done) {
        throw exception;
      }

      var context = this;
      function handle(loc, caught) {
        record.type = "throw";
        record.arg = exception;
        context.next = loc;
        return !!caught;
      }

      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        var record = entry.completion;

        if (entry.tryLoc === "root") {
          // Exception thrown outside of any try block that could handle
          // it, so set the completion value of the entire function to
          // throw the exception.
          return handle("end");
        }

        if (entry.tryLoc <= this.prev) {
          var hasCatch = hasOwn.call(entry, "catchLoc");
          var hasFinally = hasOwn.call(entry, "finallyLoc");

          if (hasCatch && hasFinally) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            } else if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else if (hasCatch) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            }

          } else if (hasFinally) {
            if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else {
            throw new Error("try statement without catch or finally");
          }
        }
      }
    },

    _findFinallyEntry: function(finallyLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc <= this.prev &&
            hasOwn.call(entry, "finallyLoc") && (
              entry.finallyLoc === finallyLoc ||
              this.prev < entry.finallyLoc)) {
          return entry;
        }
      }
    },

    abrupt: function(type, arg) {
      var entry = this._findFinallyEntry();
      var record = entry ? entry.completion : {};

      record.type = type;
      record.arg = arg;

      if (entry) {
        this.next = entry.finallyLoc;
      } else {
        this.complete(record);
      }

      return ContinueSentinel;
    },

    complete: function(record) {
      if (record.type === "throw") {
        throw record.arg;
      }

      if (record.type === "break" ||
          record.type === "continue") {
        this.next = record.arg;
      } else if (record.type === "return") {
        this.rval = record.arg;
        this.next = "end";
      }

      return ContinueSentinel;
    },

    finish: function(finallyLoc) {
      var entry = this._findFinallyEntry(finallyLoc);
      return this.complete(entry.completion);
    },

    "catch": function(tryLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc === tryLoc) {
          var record = entry.completion;
          if (record.type === "throw") {
            var thrown = record.arg;
            resetTryEntry(entry, i);
          }
          return thrown;
        }
      }

      // The context.catch method must only be called with a location
      // argument that corresponds to a known catch block.
      throw new Error("illegal catch attempt");
    },

    delegateYield: function(generator, resultName, nextLoc) {
      this.delegate = {
        generator: generator,
        resultName: resultName,
        nextLoc: nextLoc
      };

      return ContinueSentinel;
    }
  };
}).apply(this, Function("return [this, function GeneratorFunction(){}]")());

var co = require('co');
var semver = require('semver');
var debug = require('debug')('component-resolver:semver');
var Resolver = require('./resolver');

/**
 * Iterate through all the locals and dependencies to resolve semvers.
 *
 * @api private
 */

Resolver.prototype.semver = function () {
  // now all dependencies will be resolved
  // with github API calls
  this.resolvingSemver = true
  this.locals.forEach(this.resolveSemvers, this)
  this.dependencies.forEach(this.resolveSemvers, this)
}

/**
 * Iterate through all the .dependencies of the branch
 * and resolve.
 *
 * @param {Object} branch
 * @api private
 */

Resolver.prototype.resolveSemvers = function (branch) {
  var ch = this.channel.semver
  Object.keys(branch.dependencies).forEach(function (name) {
    var ref = branch.dependencies[name]
    if (typeof ref !== 'string') return
    ch.push(co(this.resolveSemver(branch, name, ref)))
  }, this)
}

/**
 * Resolve semver. First, we try locally, then we try remotely.
 *
 * @param {Object} branch
 * @param {String} name
 * @param {String} reference
 * @api private
 */

Resolver.prototype.resolveSemver = wrapGenerator.mark(function(branch, name, ref) {
  var slug, state, child, resolve;

  return wrapGenerator(function($ctx0) {
    while (1) switch ($ctx0.prev = $ctx0.next) {
    case 0:
      resolve = function resolve(child) {
        branch.dependencies[name] = child;
        if (!~child.dependents.indexOf(branch)) child.dependents.push(branch);
      };

      debug('resolving semver %s@%s', name, ref);
      slug = name + '@' + ref;
      state = this.state[slug];

      if (!(state === 'resolving')) {
        $ctx0.next = 11;
        break;
      }

      $ctx0.next = 7;
      return this.await(slug);
    case 7:
      $ctx0.t0 = $ctx0.sent;
      return $ctx0.abrupt("return", resolve($ctx0.t0));
    case 11:
      if (!(typeof state === 'object')) {
        $ctx0.next = 13;
        break;
      }

      return $ctx0.abrupt("return", resolve(state));
    case 13:
      this.state[slug] = 'resolving';
      this.status[slug] = true;
      $ctx0.t1 = this.resolveSemverLocally(branch, name, ref);

      if ($ctx0.t1) {
        $ctx0.next = 19;
        break;
      }

      return $ctx0.delegateYield(this.resolveSemverRemotely(branch, name, ref), "t2", 18);
    case 18:
      $ctx0.t1 = $ctx0.t2;
    case 19:
      ref = $ctx0.t1;
      return $ctx0.delegateYield(this.branchDependency(name, ref, branch), "t3", 21);
    case 21:
      child = $ctx0.t3;
      resolve(child);
      debug('resolved semver %s -> %s@%s', slug, name, ref);
      this.emit(slug, child);
      this.state[slug] = child;
      delete this.status[slug];
    case 27:
    case "end":
      return $ctx0.stop();
    }
  }, this);
});

/**
 * Resolve semver locally without touching the remote if possible
 *
 * @param {String} name
 * @param {String} ref
 * @return {String} reference
 * @api private
 */

Resolver.prototype.resolveSemverLocally = function (branch, name, ref) {
  var references = this.getReferences(name)
  var versions = references
    .filter(semver.valid)
    .sort(semver.rcompare)
  var version = semver.maxSatisfying(versions, ref)
  if (version) return version
  // can not be resolved with semver
  if (ref !== '*') return
  var branches = references.filter(function (ref) {
    // they need a method for this :(
    return !semver.valid(ref)
  })
  // return any release
  if (branches.length) return branches[0]
}

/**
 * Resolve semver by accessing the remote.
 *
 * @param {String} name
 * @param {String} ref
 * @return {String} reference
 * @api private
 */

Resolver.prototype.resolveSemverRemotely = wrapGenerator.mark(function(branch, name, ref) {
  var remotes, remote, versions, version;

  return wrapGenerator(function($ctx1) {
    while (1) switch ($ctx1.prev = $ctx1.next) {
    case 0:
      remotes = branch.resolvedRemotes;
    case 1:
      return $ctx1.delegateYield(this.remote.check(remotes, name), "t4", 2);
    case 2:
      remote = $ctx1.t4;

      if (!!remote) {
        $ctx1.next = 5;
        break;
      }

      throw new Error('no remote found for dependency "' + name + '".');
    case 5:
      return $ctx1.delegateYield(remote.versions(name), "t5", 6);
    case 6:
      versions = $ctx1.t5;
      version = semver.maxSatisfying(versions, ref);

      if (!version) {
        $ctx1.next = 10;
        break;
      }

      return $ctx1.abrupt("return", version);
    case 10:
      remotes = remotes.filter(nonLocal);
    case 11:
      if (remote.name === 'local' && ref !== '*') {
        $ctx1.next = 1;
        break;
      }
    case 12:
      if (!(ref !== '*')) {
        $ctx1.next = 14;
        break;
      }

      throw new Error('invalid semver range "' + ref + '" for "' + name + '"');
    case 14:
      return $ctx1.abrupt("return", 'master');
    case 15:
    case "end":
      return $ctx1.stop();
    }
  }, this);
});

function nonLocal(remote) {
  return (remote !== 'local');
}
